source/extensions/tracers/opencensus/opencensus_tracer_impl.cc

```diff
@@ -0,0 +1,169 @@
+#include "extensions/tracers/opencensus/opencensus_tracer_impl.h"
+
+#include "absl/strings/str_cat.h"
+#include "opencensus/exporters/trace/stackdriver/stackdriver_exporter.h"
+#include "opencensus/exporters/trace/stdout/stdout_exporter.h"
+#include "opencensus/exporters/trace/zipkin/zipkin_exporter.h"
+#include "opencensus/trace/propagation/cloud_trace_context.h"
+#include "opencensus/trace/sampler.h"
+#include "opencensus/trace/span.h"
+#include "opencensus/trace/trace_config.h"
+#include "opencensus/trace/trace_params.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace OpenCensus {
+
+namespace {
+
+/**
+ * OpenCensus tracing implementation of the Envoy Span object.
+ */
+class Span : public Tracing::Span {
+public:
+  Span(const Tracing::Config& config, const envoy::config::trace::v2::OpenCensusConfig* oc_config,
+       Http::HeaderMap& request_headers, const std::string& operation_name, SystemTime start_time,
+       const Tracing::Decision tracing_decision);
+
+  // Used by spawnChild().
+  Span(const envoy::config::trace::v2::OpenCensusConfig* oc_config,
+       ::opencensus::trace::Span&& span);
+
+  void setOperation(const std::string& operation) override;
+  void setTag(const std::string& name, const std::string& value) override;
+  void finishSpan() override;
+  void injectContext(Http::HeaderMap& request_headers) override;
+  Tracing::SpanPtr spawnChild(const Tracing::Config& config, const std::string& name,
+                              SystemTime start_time) override;
+  void setSampled(bool sampled) override;
+
+private:
+  // TODO: Do we take the sampling decision from StartSpan, or use the sampler
+  // configured in TraceConfig?
+  ::opencensus::trace::Sampler* GetSampler(bool is_sampled) {
+    static ::opencensus::trace::AlwaysSampler always_sampler;
```

Comment:
These two samplers have no fields and their constructors and destructors are no-ops. This usage should be safe - what's the risk here?

Reply:
This is now enforced by clang-tidy, so it must be fixed :)

Reply:
Hmm, we might not have turned that on yet (@lizan I thought we had?). In any case, it's probably safe in this case, but we have a strong preference to avoid this C++ anti-pattern, having been burnt by it in the past. The fact that you have to understand abstraction breaking details in order to know whether a line of code is safe is as static is not a great start. The safe way to do this in Envoy is with `CONSTRUCT_ON_FIRST_USE`, lots of examples of this.

Reply:
Fair point. I've made them non-static.